Python: Extended Write Example

ax3l opened the pull request:
Add the `7_extended_write_serial` example in python.

ax3l committed:
Python: Extended Write Example

Add the `7_extended_write_serial` example in python.

ax3l commented on src/binding/python/PatchRecordComponent.cpp:
```diff
@@ -0,0 +1,47 @@
+        .def_property_readonly("shape", &PatchRecordComponent::getExtent)
+
+        /*
+        template< typename T >
```
todo: implement

anokfireball commented on examples/7_extended_write_serial.py:
```diff
@@ -0,0 +1,195 @@
+        .set_time_unit_SI(1.39e-16)
+    # everything that is accessed with [] should be interpreted as permanent
+    # storage the objects sunk into these locations are deep copies
+    del tmpItObj
```
Is this `del` supposed to signal that only the *handle* and not the underlying sunk resource is deleted?
If so, a comment to explain what is going on should be added.

ax3l replied:
Let's check this again when updating the C++ example

ax3l replied:
probably not needed anymore. I did not control the garbage collector properly before `0.6.0-alpha`

anokfireball commented on examples/7_extended_write_serial.py:
```diff
@@ -0,0 +1,195 @@
+
+    # alternatively, a copy may be created and later re-assigned to
+    # f.iterations[1]
+    copy = f.iterations[1]  # TODO .copy()
```
This is actually outdated in the C++ example. Please add the **TODO** there as well.

ax3l replied:
Yes, I thought so. Let's update both.

anokfireball commented on examples/7_extended_write_serial.py:
```diff
@@ -0,0 +1,195 @@
+    cur_it.meshes["generic_2D_field"].set_unit_dimension(
+        {Unit_Dimension.L: -3, Unit_Dimension.M: 1})
+
+    # as this is a reference, it modifies the original resource
```
Outdated in the C++ example as well.

anokfireball commented on examples/7_extended_write_serial.py:
```diff
@@ -0,0 +1,195 @@
+
+    # as this is a reference, it modifies the original resource
+    lowRez = cur_it.meshes["generic_2D_field"]
+    lowRez.set_grid_spacing([1, 1, 1, 1, 1, 1]) \
```
This `gridSpacing` would actually make it a 6D field.

ax3l replied:
huh, don't we have the same in the C++ example? Should be three values in both examples, no?

anokfireball replied:
The [C++ code](https://github.com/openPMD/openPMD-api/blob/0.6.0-alpha/examples/7_extended_write_serial.cpp#L98) for this reads
```cpp
/* construct a vector of [6, 1] */
lowRez.setGridSpacing(std::vector< double >{6, 1}).setGridGlobalOffset({0, 600});
```
Your interpretation was
```cpp
/* construct a vector of six 1's */
lowRez.setGridSpacing(std::vector< double >(6, 1)).setGridGlobalOffset({0, 600});
```

> Should be three values in both examples, no?

As this is a **2D** mesh, it should contain two values: `[6, 1]`.

https://github.com/openPMD/openPMD-standard/blob/1.1.0/STANDARD.md#required-attributes-for-each-mesh-record

> `gridSpacing`
type: 1-dimensional array containing N *(floatX)* elements, where N is the **number of dimensions in the simulation**

ax3l replied:
Oh wow, that's dangerous xD
I thought the initializer-list constructor should read `({6, 1})` but your code is indeed ok and does exactly what you say.

anokfireball reviewed (approved):
Python example looks good, but C++ needs some love. Will do that when tackling #171.

ax3l commented:
Ah cheese, I should remove Python 2 builds altogether.

ax3l committed:
Patch Example: Py2 Compatible

Since in Python 2, numpy scalars are not properly matched
to the pybind11::buffer overloads, we write a simple array
in patch store.

ax3l committed:
make_constant, store: 1-element arrays

For user convienience, allow also n-dimensional arrays when numpy
scalar values are requested.

ax3l committed:
Changelog: Extended Py Write Example
